electron/process_manager.ts

```diff
@@ -71,39 +73,66 @@ const UDP_FORWARDING_TEST_RETRY_INTERVAL_MS = 1000;
 //
 // This function is roughly the Electron counterpart of Android's VpnTunnelService.startShadowsocks.
 export function startVpn(config: cordova.plugins.outline.ServerConfig, onDisconnected: () => void) {
-  return isServerReachable(config)
-      .then(() => {
-        return startLocalShadowsocksProxy(config, onDisconnected);
-      })
-      .then(() => {
-        return validateServerCredentials();
-      })
-      .then(() => {
-        return checkUdpForwardingEnabled();
-      })
-      .then(() => {
-        return startTun2socks(config.host || '', onDisconnected);
-      })
-      .then(() => {
-        return routingService.configureRouting(TUN2SOCKS_VIRTUAL_ROUTER_IP, config.host || '');
-      })
-      .catch((e) => {
-        stopProcesses();
-        throw e;
-      });
+  // RoutingService requires an IP; perform a lookup now and use that throughout setup.
+  return lookup(config.host || '').then((ip: string) => {
+    return isServerReachableByIp(ip, config.port || 0)
+        .then(() => {
+          return startLocalShadowsocksProxy(ip, config, onDisconnected);
+        })
+        .then(() => {
+          return validateServerCredentials();
+        })
+        .then(() => {
+          return checkUdpForwardingEnabled();
+        })
+        .then(() => {
+          return startTun2socks(ip, onDisconnected);
+        })
+        .then(() => {
+          return routingService.configureRouting(TUN2SOCKS_VIRTUAL_ROUTER_IP, ip);
+        })
+        .catch((e) => {
+          stopProcesses();
+          throw e;
+        });
+  });
+}
+
+// Uses the OS' built-in functions, i.e. /etc/hosts, et al.:
+// https://nodejs.org/dist/latest-v10.x/docs/api/dns.html#dns_dns
+//
+// Effectively a no-op if hostname is already an IP.
+function lookup(hostname: string): Promise<string> {
```

Comment:
nit: maybe rename to `lookupIp` or `lookupIpAddress`?

Reply:
Done.